pandas/core/common.py

```diff
@@ -381,25 +382,17 @@ def _asarray_tuplesafe(values, dtype=None):
         return values.values
 
     if isinstance(values, list) and dtype in [np.object_, object]:
-        return lib.list_to_object_array(values)
+        return construct_1d_array_from_listlike(values)
 
     result = np.asarray(values, dtype=dtype)
 
     if issubclass(result.dtype.type, compat.string_types):
         result = np.asarray(values, dtype=object)
 
     if result.ndim == 2:
-        if isinstance(values, list):
-            return lib.list_to_object_array(values)
-        else:
-            # Making a 1D array that safely contains tuples is a bit tricky
-            # in numpy, leading to the following
-            try:
-                result = np.empty(len(values), dtype=object)
-                result[:] = values
-            except ValueError:
-                # we have a list-of-list
-                result[:] = [tuple(x) for x in values]
+        if hasattr(values, '__array__'):
+            values = [tuple(x) for x in values]
```

Comment:
I don't understand this one (can you add some comments?)  
The original comment said "we have a list-of-list", which is actually not possible because it was already checked before if it was a list.

Reply:
Yes, that comment was wrong, and confusing. Will add a comment.